### ISSUE TITLE
LinearRegressor with OrdinaryLeastSquares

### ISSUE THREAD

**Issue by WaylonBrown** [gyrdym/ml_algo#198]
I see we can only choose between Gradient and Coordinate. I believe this is why the data I'm getting back isn't what I expect. If that's not the reason, please let me know, thanks!

For example, if I have the data

```
Grind, RoastLevel, Time
5.5, 5, 20
5.25, 5, 22
```

And I make RoastLevel and Time independent variables, and Grind dependent, I can get an expected prediction with Python using Sklearn out of the box.

````
data_minus_grind = []
grind_data = []
for row in floats:
    data_minus_grind.append(row[1:3])
    grind_data.append(row[0])

model = LinearRegression().fit(data_minus_grind, grind_data)
prediction_data = [[5,25]]
prediction = model.predict(prediction_data)
````

This prediction, trying to predict the Grind for the Time **25**, I receive **4.875**, which sounds about right. Grind should go down while Time goes up.

However, if I try to use this library, my slope is always in the wrong direction, with the two variables moving in the same direction for some reason.

If I try

```
final samples = DataFrame.fromRawCsv(rawCsvContent, headerExists: true);
const targetColName = "Grind";
final defaultRegressor = LinearRegressor(
    samples,
    targetColName
);
final dataToPredict = [
    // Roast level, time
    [ 5, 25.0 ]
];
final dataframeToPredict = DataFrame(dataToPredict, headerExists: false);
final prediction = regressor.predict(dataframeToPredict);
```

Then the result for Grind with a Time of 25.0 is **6.79**. As I move up the time, Grind should decrease, but instead it increases. I've tried tweaking many of the parameters but haven't found a fix.

Thanks!

**Comment by gyrdym** [comment]
@WaylonBrown Thank you for creating the issue!

Yes, the problem is in the method which is used by LinearRegressor. As far as I understand, the ordinary least squares algorithm in sklearn is based on [singular value decomposition](https://scikit-learn.org/stable/modules/linear_model.html#ordinary-least-squares-complexity), which gives different results. Apparently, I should consider adding something similar to the library :)

**Comment by WaylonBrown** [comment]
Awesome, thanks! I take it that implementation would take a while, in the meantime are there any parameters I can tweak with this library to get what I want? I'd like to use it if possible.

**Comment by gyrdym** [comment]
@WaylonBrown I'm afraid, there are no such parameters. Yes, it takes some time to implement something similar to sklearn, unfortunately.

**Comment by gyrdym** [comment]
@WaylonBrown Hi!

If it's still relevant for you - I added the closed-form solution of the linear regression problem (ml_algo version 16.3.0), now linear regression performs as it does in sklearn.

You can try the same code as you tried before:

```dart
final samples = DataFrame.fromRawCsv(rawCsvContent, headerExists: true);
const targetColName = "Grind";
final defaultRegressor = LinearRegressor(
    samples,
    targetColName
);
final dataToPredict = [
    // Roast level, time
    [ 5, 25.0 ]
];
final dataframeToPredict = DataFrame(dataToPredict, headerExists: false);
final prediction = regressor.predict(dataframeToPredict);
```

The prediction will be approzimately `5`

**Comment by WaylonBrown** [comment]
That is very awesome of you to add the quick update, thank you gyrdym! I'll use it ASAP :)

**Comment by gyrdym** [comment]
@WaylonBrown you're welcome)!
